### PR TITLE
prevent revisiting slots in get_closest_completion

### PR DESCRIPTION
**Pull request by jbiseda**
#### Problem
`get_closest_completion` does not check `processed_slots` when following parent slots.

#### Summary of Changes
- Only insert slots into `processed_slots` when calling `generate_repairs_for_slot`.
- use separate HashSet for slot tracking during closeness calculation
- skip already processed slots before generating repairs

Fixes #

### CODE REVIEW

**Review by bw-solana — APPROVED** [review]
LGTM

File: core/src/repair_generic_traversal.rs [hunk 1]
```diff
@@ -116,12 +116,14 @@ pub fn get_closest_completion(
     processed_slots: &mut HashSet<Slot>,
     limit: usize,
 ) -> (Vec<ShredRepairType>, /* processed slots */ usize) {
-    let mut v: Vec<(Slot, u64)> = Vec::default();
+    let mut slot_dists: Vec<(Slot, u64)> = Vec::default();
+    let mut explored_slots: HashSet<Slot> = HashSet::default();
     let iter = GenericTraversal::new(tree);
     for slot in iter {
-        if processed_slots.contains(&slot) {
+        if processed_slots.contains(&slot) || explored_slots.contains(&slot) {
```
Comment (behzadnouri) [inline core/src/repair_generic_traversal.rs:123]: Doesn't `GenericTraversal` here actually iterate over unique slots?

Comment (jbiseda) [reply core/src/repair_generic_traversal.rs:123]: Doesn't HeaviestSubtreeForkChoice allow duplicate slots?
https://github.com/solana-labs/solana/blob/5aa5aef1a9682aff63a1d13222b2f2b24501021c/core/src/heaviest_subtree_fork_choice.rs#LL3820C8-L3820C29

Comment (jbiseda) [reply core/src/repair_generic_traversal.rs:123]: > Doesn't `GenericTraversal` here actually iterate over unique slots?
Yeah, there's no hash distinction: see #30516

**Pull request merged**